text/WildcardMatching/Wildcard Matching - Kamil Rajtar.tex

```diff
@@ -0,0 +1,64 @@
+\documentclass{article}
+\usepackage[utf8]{inputenc}
+
+\title{Wyszukiwanie z wildcardami w tekście i wzorcu}
+\author{Kamil Rajtar}
+\date{Czerwiec 2020}
+
+\begin{document}
+
+\newcommand\defeq{\stackrel{\mathclap{\normalfont\mbox{def}}}{=}}
+
+\maketitle
+
+\section*{Opis problemu}
+
+Na wejściu do algorytmu dostajemy tekst i wzorzec w których oprócz liter mogą występować wildcardy (?) pasujące do dowolnego znaku. Na wyjściu ma się znaleźć lista pozycji w których wzorzec pasuje do tekstu.
+
+\section*{Wersja bez wildcardów}
+
+W rozwiązaniu problemu użyty jest splot obliczany za pomocą szybkiej transformaty Furiera (FFT) zdefiniowany następująco:
+
+$$
+    p \otimes t \defeq (\sum^{m-1}_{j=0}p_{j}t_{i+j}, 0 \leq i \leq n-m)
+$$
+
+Normalne (bez wildcardów) dopasowanie za pomocą splotu tekstu (t) i odwróconego wzorca (p) możemy obliczyć za pomocą wzoru:
+
+$$
+    \sum_{j-0}^{m-1} (p_j - t_{i+j})^2 = \sum_{j-0}^{m-1} (p^2_{j} - 2 p_{j}t_{i+j}+t^2_{i+j})
+$$
+
+Zobaczmy że lewa część równania równa 0 to znaleźliśmy dopasowanie. (Tekst nie różni się od wzorca na żadnej z m kolejnych pozycji). Prawą stronę potrafimy szybko obliczyć ponieważ podniesienie każdej pozycji do kwadratu wykonywane jest w czasie stałym a środkowy składnik liczmy za pomocą FFT w O(n log m).
```

Comment:
1. There are some problems with the syntax
2. Remove parentheses
3. Use math mode, whenever necessary.

Additionally, include a simple proof in both ways that $LHS = 0$ is indeed equivalent to match.

Moreover, $O(n \log{m})$ time for the middle part may be indeed simple, but it's definitely non-trivial:
 - if we were to compute it from scratch for every $i$ we would have $O(m \log{m})$ for each so $O(n m \log{m})$ in total,
 - if we were to compute it for the whole $t$ and $p$ once then we get $O(n \log{n})$, as we need to pad $p$.

Please explain it further.